File: packages/core/src/ratelimit/lua/leakyBucket.lua
-- KEYS[1] = queue key
-- ARGV[1] = limit, ARGV[2] = windowMs, ARGV[3] = capacity, ARGV[4] = now, ARGV[5] = cost
local limit = tonumber(ARGV[1])
local windowMs = tonumber(ARGV[2])
local capacity = tonumber(ARGV[3])
local now = tonumber(ARGV[4])
local cost = tonumber(ARGV[5])
local leakPerMs = limit / windowMs

local raw = redis.call('HMGET', KEYS[1], 'level', 'ts')
local level = tonumber(raw[1]) or 0
local lastLeakMs = tonumber(raw[2]) or now

local elapsed = math.max(0, now - lastLeakMs)
level = math.max(0, level - elapsed * leakPerMs)

local allowed = (level + cost) <= capacity
if allowed then
  level = level + cost
end

redis.call('HSET', KEYS[1], 'level', level, 'ts', now)
redis.call('PEXPIRE', KEYS[1], math.ceil(capacity / leakPerMs) + 1000)

local retryAfter = 0
if not allowed then
  retryAfter = math.ceil((level + cost - capacity) / leakPerMs)
end

return { allowed and 1 or 0, math.max(0, math.floor(capacity - level)), retryAfter }

File: packages/core/src/ratelimit/algorithms/leakyBucket.ts
import type { Policy } from '../stores/Store.js';
import type { AlgorithmOutput } from './types.js';

export interface LeakyBucketState {
  readonly level: number;
  readonly lastLeakMs: number;
}

/**
 * Sabit hızda boşalan kuyruk. Çıkışı düzleştirir ama burst'ü emmez
 * (bkz. PLAN.md §4).
 */
export function leakyBucket(
  state: LeakyBucketState | undefined,
  policy: Policy,
  now: number,
  cost: number,
): AlgorithmOutput<LeakyBucketState> {
  const capacity = policy.burst ?? policy.limit;
  const leakPerMs = policy.limit / policy.windowMs;
  const lastLeakMs = state?.lastLeakMs ?? now;
  const elapsed = Math.max(0, now - lastLeakMs);
  const level = Math.max(0, (state?.level ?? 0) - elapsed * leakPerMs);

  const allowed = level + cost <= capacity;
  const nextLevel = allowed ? level + cost : level;

  return {
    state: { level: nextLevel, lastLeakMs: now },
    result: {
      allowed,
      limit: policy.limit,
      remaining: Math.max(0, Math.floor(capacity - nextLevel)),
      retryAfterMs: allowed ? 0 : Math.ceil((level + cost - capacity) / leakPerMs),
    },
  };
}
